### DISCUSSION TITLE
Gitian builder

### DISCUSSION THREAD

**Discussion by micaelmalta** [dogecoin/dogecoin#2394] [Show and tell]
Hello everyone, as a new joiner to the community, it has been very hard to follow the Gitian doc to build and sign builds

I worked on a new repo that make easy to anyone to share their Gitian sha256sum

https://github.com/micaelmalta/gitian-builder-dogecoin

I also had to change the gitian-builder based on @patricklodder work

https://github.com/micaelmalta/gitian-builder

Can we consider to integrate it in the Official Dogecoin Universe so many developers can use it and validate our builds ?

Thanks @AbcSxyZ for testing it ;)

Micael

**Comment by micaelmalta** [comment]
I integrated the Gitian Builder on the CI and everything is passing:

https://github.com/micaelmalta/dogecoin/pull/13
https://cirrus-ci.com/build/4744566927523840

**Comment by AbcSxyZ** [comment]
Nice to see a working Gitian :) How do you want to integrate it to the Official Dogecoin Universe ?

It would be great to have working script within the main repo and avoid relying on external one, are you thinking about doing PRs for it ? That's for me the best integration we could have.
I guess, at least for the possibility you would have to do an unverified change on your repo to compromise all executables of an entire release.

About [gitian-builder](https://github.com/micaelmalta/gitian-builder), can you tell me briefly what is the issue with `apt-cacher` and [fixes](https://github.com/micaelmalta/gitian-builder/commits/docker) you made ? I saw you disabled it and add an option for, is there a bug or do we need a specific setup for dogecoin ? The [devrandom repo](https://github.com/devrandom/gitian-builder) is active, wondering if a fork is really needed or if a fix to their repo might be something to think about. Would be great to avoid a fork, otherwise it needs some maintenance...

**Reply by micaelmalta** [reply]
@AbcSxyZ a PR is open on devrandom repo so like that we dont need a fork : https://github.com/devrandom/gitian-builder/pull/249

IMO, Having to use apt-cacher by default is overkill and need extra prerequisite steps to make it work. I added an option to enable it. Also, apt-cacher is not the most stable software ...

**Reply by AbcSxyZ** [reply]
Great, it may be a matter of time to have the change incorporated, maintainers look interested.

**Reply by AbcSxyZ** [reply]
And for your script, are you thinking about doing PR to the Dogecoin repository ? Do you agree with the idea that we can't rely on an external repository managed by an individual for the gitian process ?

**Reply by micaelmalta** [reply]
I totally agree! I would suggest to integrate the gitian-builder-dogecoin repo into https://github.com/dogecoin/

**Comment by patricklodder** [comment]
First off, thank you for working on this! Looking awesome & pretty!

I was thinking, why wouldn't we put this under `./contrib`?

**Reply by micaelmalta** [reply]
@patricklodder You're welcome!

I was thinking on an independent Github repo for this reason: The script need to clone the Dogecoin repo even though you are already on it (sic). It makes no sense to do it :)

**Reply by patricklodder** [reply]
because of the descriptors?

**Reply by micaelmalta** [reply]
because you need to clone the repo to get the branch/commit you want to build.

We could do that on the current dogecoin folder you are in but it would be possible only to build the branch you are already on. Otherwise we would have to reset it...

**Reply by micaelmalta** [reply]
@patricklodder  And also because of the descriptors yes

If we move the descriptors to the gitian-builder repo, it would be all good (I think!)

**Reply by micaelmalta** [reply]
@patricklodder  Here would be the tree during the build

```
-dogecoin
    - contrib
    - src
    - gitian-builder
    - gitian.sigs
    - dogecoin-detached-sigs
    - dogecoin-binaries
    - dogecoin
        - contrib
        - src 
```

**Comment by micaelmalta** [comment]
Here is the final repo working: https://github.com/micaelmalta/gitian-builder-dogecoin

**Comment by micaelmalta** [comment]
Hello @patricklodder @AbcSxyZ !

A new version has been pushed with a nice feature: You can now build any GITIAN repo/branch on Cirrus!

DEMO: https://cirrus-ci.com/build/4704793919750144

HOW TO:
1/ Fork repo: https://github.com/micaelmalta/gitian-builder-dogecoin
2/ Git clone
3/ git checkout -b build_**whatever**   # IMPORTANT PART AS GITIAN BUILD ARE ONLY BUILD ON `build_*` branches
4/ vim .cirrus.yml
5/ Change those 2 values: 
```  
URL: https://github.com/mmicael/dogecoin
BRANCH: 1.14.4-libevent
```
6/ git commit -m "whatever"
7/ git push
8/ Wait 45mn!
9/ ENJOY!

**Comment by micaelmalta** [comment]
@AbcSxyZ The script is not dependent of a fork anymore; instead It's downloading a fixed commit from the original repo and apply a fix: https://github.com/micaelmalta/gitian-builder-dogecoin/blob/master/gitian-build.sh#L261

@patricklodder Descriptors are not an issue anymore as the script download them directly now: https://github.com/micaelmalta/gitian-builder-dogecoin/blob/master/gitian-build.sh#L271

**Comment by micaelmalta** [comment]
PR integrating the new Gitian Builder into DOGECOIN: https://github.com/dogecoin/dogecoin/pull/2396 🚀
